src/Dictionaries/ExternalQueryBuilder.cpp

```diff
@@ -152,95 +165,135 @@ void ExternalQueryBuilder::composeLoadAllQuery(WriteBuffer & out) const
+
+        auto query_copy = query;
+        query_copy.replace(condition_position, condition_value.size(), condition_value);
+
+        return query_copy;
+    }
 }
 
 
-std::string ExternalQueryBuilder::composeLoadIdsQuery(const std::vector<UInt64> & ids)
+std::string ExternalQueryBuilder::composeLoadIdsQuery(const std::vector<UInt64> & ids) const
 {
     if (!dict_struct.id)
         throw Exception(ErrorCodes::UNSUPPORTED_METHOD, "Simple key required for method");
 
     WriteBufferFromOwnString out;
-    writeString("SELECT ", out);
 
-    if (!dict_struct.id->expression.empty())
+    if (query.empty())
     {
-        writeParenthesisedString(dict_struct.id->expression, out);
-        writeString(" AS ", out);
-    }
+        writeString("SELECT ", out);
 
-    writeQuoted(dict_struct.id->name, out);
+        if (!dict_struct.id->expression.empty())
+        {
+            writeParenthesisedString(dict_struct.id->expression, out);
+            writeString(" AS ", out);
+        }
 
-    for (const auto & attr : dict_struct.attributes)
-    {
-        writeString(", ", out);
+        writeQuoted(dict_struct.id->name, out);
 
-        if (!attr.expression.empty())
+        for (const auto & attr : dict_struct.attributes)
         {
-            writeParenthesisedString(attr.expression, out);
-            writeString(" AS ", out);
+            writeString(", ", out);
+
+            if (!attr.expression.empty())
+            {
+                writeParenthesisedString(attr.expression, out);
+                writeString(" AS ", out);
+            }
+
+            writeQuoted(attr.name, out);
         }
 
-        writeQuoted(attr.name, out);
-    }
+        writeString(" FROM ", out);
+        if (!db.empty())
+        {
+            writeQuoted(db, out);
+            writeChar('.', out);
+        }
+        if (!schema.empty())
+        {
+            writeQuoted(schema, out);
+            writeChar('.', out);
+        }
 
-    writeString(" FROM ", out);
-    if (!db.empty())
-    {
-        writeQuoted(db, out);
-        writeChar('.', out);
-    }
-    if (!schema.empty())
-    {
-        writeQuoted(schema, out);
-        writeChar('.', out);
-    }
+        writeQuoted(table, out);
 
-    writeQuoted(table, out);
+        writeString(" WHERE ", out);
 
-    writeString(" WHERE ", out);
+        if (!where.empty())
+        {
+            writeString(where, out);
+            writeString(" AND ", out);
+        }
 
-    if (!where.empty())
-    {
-        writeString(where, out);
-        writeString(" AND ", out);
+        composeIdsCondition(ids, out);
+        writeString(";", out);
+
+        return out.str();
     }
+    else
+    {
+        writeString(query, out);
 
-    writeQuoted(dict_struct.id->name, out);
-    writeString(" IN (", out);
+        auto condition_position = query.find("{condition}");
+        if (condition_position == std::string::npos)
+        {
```

Comment:
@vdimir this option is for real clickhouse experts.
If {condition} is specified then it should be part of WHERE section and condition will be filled with additional predicate.
If {condition} is not specified then WHERE is added to the end of the query to support simple case.

Also WHERE section is not supported with custom query.

Reply:
Will it substitute {condition} with something reasonable for non cache-a like dictionaries?

Reply:
@filimonov {condition} is necessary only if your dictionary load keys on requests like cache, direct.
Or if client uses update_field then condition will be filled with `update_field >= last_update_time`.

Reply:
Maybe it would be simpler to replace it with `1` in all other cases (if it appears).

Reply:
@filimonov it seems that if user is using this setting he should be aware a bit how dictionaries internals work. Possibly it can be replaced by 1 in case if it is presented, but probably better to show error to user.